@ariakit/test: press Shift+Home / Shift+End ignore selectionDirection, extending the selection past the anchor

### Current behavior

`@ariakit/test`'s `press` simulates the browser's default keydown behavior for navigation keys in text fields, and its doc comment explicitly advertises "moving the caret with the arrow and `Home`/`End` keys" ([press.ts#L262-L287](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L262-L287)). But the `Home` and `End` handlers pick the fixed endpoint of the new selection without looking at `selectionDirection` ([press.ts#L100-L114](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L100-L114)):

```ts
async Home(element, { shiftKey }) {
  if (isTextField(element)) {
    const { value, selectionEnd } = element;
    const end = Math.min(value.length, shiftKey ? (selectionEnd ?? 0) : 0);
    setSelectionRange(element, 0, end, "backward");
  }
},

async End(element, { shiftKey }) {
  if (isTextField(element)) {
    const { value, selectionStart } = element;
    const start = shiftKey ? (selectionStart ?? 0) : value.length;
    setSelectionRange(element, start, value.length, "forward");
  }
},
```

In real browsers, Shift+Home and Shift+End move the selection *focus* while the *anchor* stays put. For a forward selection the anchor is `selectionStart`; for a backward selection it is `selectionEnd`. So with `value = "hello world"` and a forward selection 6–8 (anchor 6), Shift+Home produces 0–6 backward — the portion past the anchor is deselected. This code instead keeps `selectionEnd` (8) and produces 0–8. Symmetrically, with a backward selection 4–6 (anchor 6), Shift+End produces 6–11 forward, but this code keeps `selectionStart` (4) and produces 4–11.

The adjacent `ArrowUp`/`ArrowDown` handlers get this exact anchor logic right by consulting `selectionDirection` ([press.ts#L150-L184](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L150-L184)), which shows the intended fidelity and confirms the property is available in the supported test environments (happy-dom and jsdom both implement it). No guard, comment, or test compensates for the divergence in `Home`/`End`; within the repo no test exercises Shift+Home or Shift+End at all (the only shifted selection key in use is `Shift+ArrowLeft` in [`examples/combobox-group/test.ts#L65-L67`](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/examples/combobox-group/test.ts#L65-L67), which is unaffected).

A consumer testing selection-dependent behavior — e.g. combobox inline autocomplete, where the completed suffix is held as a backward/forward selection — gets a simulated selection no browser would produce. The failure mode is quiet: an assertion on `selectionStart`/`selectionEnd` after Shift+Home/End either fails for the wrong reason or, worse, passes against behavior that diverges from what users see.

The same user scenario was verified in real Chromium with Playwright (real key presses only): a forward 6–8 selection becomes 0–6 backward under Shift+Home, and a backward 4–6 selection becomes 6–11 forward under Shift+End — the anchor stays fixed in both cases.

### Steps to reproduce the bug

For a testing library, running a test is the consumer reproduction. The minimal repro is the following test file, which asserts the real-browser values and therefore fails on current `@ariakit/test`. In the Ariakit repo, drop it in `app/src/sandbox/test-shift-home-end/` next to this component:

```tsx
// index.react.tsx
import type { SyntheticEvent } from "react";
import { useState } from "react";

export default function Example() {
  const [start, setStart] = useState<number | null>(null);
  const [end, setEnd] = useState<number | null>(null);
  const [direction, setDirection] = useState<string | null>(null);

  function readSelection(event: SyntheticEvent<HTMLInputElement>) {
    const input = event.currentTarget;
    setStart(input.selectionStart);
    setEnd(input.selectionEnd);
    setDirection(input.selectionDirection);
  }

  return (
    <div>
      <label htmlFor="field">Field</label>
      <input
        id="field"
        defaultValue="hello world"
        onSelect={readSelection}
        onKeyUp={readSelection}
        onMouseUp={readSelection}
      />
      <p>
        <output aria-label="selectionStart">{String(start)}</output>{" "}
        <output aria-label="selectionEnd">{String(end)}</output>{" "}
        <output aria-label="selectionDirection">{String(direction)}</output>
      </p>
    </div>
  );
}
```

```ts
// test.ts
import { focus, press, q } from "@ariakit/test";
import { expect, test } from "vitest";

function getField() {
  return q.textbox.ensure("Field") as HTMLInputElement;
}

test("press.Home with shiftKey keeps the anchor of a forward selection", async () => {
  const field = getField();
  await focus(field);
  field.setSelectionRange(6, 8, "forward"); // forward selection 6-8, anchor 6

  await press.Home(field, { shiftKey: true });

  // Real browsers move the focus to the start while the anchor (6) stays put.
  expect(field.selectionStart).toBe(0);
  expect(field.selectionEnd).toBe(6); // current @ariakit/test yields 8
  expect(field.selectionDirection).toBe("backward");
});

test("press.End with shiftKey keeps the anchor of a backward selection", async () => {
  const field = getField();
  await focus(field);
  field.setSelectionRange(4, 6, "backward"); // backward selection 4-6, anchor 6

  await press.End(field, { shiftKey: true });

  // Real browsers move the focus to the end while the anchor (6) stays put.
  expect(field.selectionStart).toBe(6); // current @ariakit/test yields 4
  expect(field.selectionEnd).toBe(11);
  expect(field.selectionDirection).toBe("forward");
});
```

Run `pnpm test-react test-shift-home-end` from the repo root. Both tests fail:

```
AssertionError: expected 8 to be 6   // Shift+Home keeps the old focus end 8
AssertionError: expected 4 to be 6   // Shift+End keeps the old focus end 4
```

The exact same user scenario was verified in a real browser with Playwright (`test-browser.ts`, real `Home`/`ArrowRight`/`Shift+ArrowLeft`/`Shift+Home`/`Shift+End` key presses on the rendered input): both tests pass in Chromium, which keeps the anchor fixed (0–6 backward and 6–11 forward).

To reproduce on StackBlitz with the released package, open https://stackblitz.com/github/ariakit/ariakit/tree/main/templates/react?file=src%2Fapp.tsx, replace `src/app.tsx` with the component below, add a sibling test that runs the two `press.Home`/`press.End` cases above, and run it:

```tsx
import type { SyntheticEvent } from "react";
import { useState } from "react";

export default function App() {
  const [start, setStart] = useState<number | null>(null);
  const [end, setEnd] = useState<number | null>(null);
  const [direction, setDirection] = useState<string | null>(null);

  function readSelection(event: SyntheticEvent<HTMLInputElement>) {
    const input = event.currentTarget;
    setStart(input.selectionStart);
    setEnd(input.selectionEnd);
    setDirection(input.selectionDirection);
  }

  return (
    <div>
      <label htmlFor="field">Field</label>
      <input
        id="field"
        defaultValue="hello world"
        onSelect={readSelection}
        onKeyUp={readSelection}
        onMouseUp={readSelection}
      />
      <p>
        <output aria-label="selectionStart">{String(start)}</output>{" "}
        <output aria-label="selectionEnd">{String(end)}</output>{" "}
        <output aria-label="selectionDirection">{String(direction)}</output>
      </p>
    </div>
  );
}
```

### Expected behavior

Shift+Home and Shift+End should hold the selection anchor fixed and move only the focus endpoint, matching real browsers, jsdom, and the package's own anchor-aware `ArrowUp`/`ArrowDown` handlers. For a forward selection 6–8, Shift+Home should yield 0–6 backward; for a backward selection 4–6, Shift+End should yield 6–11 forward. The anchor is `selectionStart` for a forward selection and `selectionEnd` for a backward one; a `"none"`/`null` direction should fall back to preserving the existing endpoint, matching the current neutral case and the arrow handlers.

### Workaround

Consumer-side, in test code only. Replicate the browser's anchor-aware default action with the standard `setSelectionRange` DOM API instead of relying on `press.Home`/`press.End`, deriving the anchor from `selectionDirection`, then dispatch a keydown/keyup so any `onSelect`/`onKeyUp` mirror updates as it would after a real keypress:

```ts
import { dispatch } from "@ariakit/test";

async function shiftHome(field: HTMLInputElement) {
  const anchor =
    field.selectionDirection === "forward"
      ? (field.selectionStart ?? 0)
      : (field.selectionEnd ?? 0);
  field.setSelectionRange(0, anchor, "backward");
  await dispatch.keyDown(field, { key: "Home", shiftKey: true });
  await dispatch.keyUp(field, { key: "Home", shiftKey: true });
}

async function shiftEnd(field: HTMLInputElement) {
  const anchor =
    field.selectionDirection === "backward"
      ? (field.selectionEnd ?? 0)
      : (field.selectionStart ?? 0);
  field.setSelectionRange(anchor, field.value.length, "forward");
  await dispatch.keyDown(field, { key: "End", shiftKey: true });
  await dispatch.keyUp(field, { key: "End", shiftKey: true });
}
```

Verified: with `shiftHome`/`shiftEnd` in place of `press.Home`/`press.End`, both tests above pass on the current `@ariakit/test` (2/2 green).

### Possible solutions

Make `Home` and `End` anchor-aware the same way `ArrowUp`/`ArrowDown` already are: derive the anchor from `selectionDirection`, then move the focus endpoint to the start (Home) or end (End) of the value. The non-shift paths stay behaviorally identical (direction on a collapsed range is meaningless, and the arrow handlers already call `setSelectionRange` without one); the dropped `Math.min` clamp is redundant since `selectionStart`/`selectionEnd` never exceed `value.length`.

```diff
   async Home(element, { shiftKey }) {
-    if (isTextField(element)) {
-      const { value, selectionEnd } = element;
-      const end = Math.min(value.length, shiftKey ? (selectionEnd ?? 0) : 0);
-      setSelectionRange(element, 0, end, "backward");
-    }
+    if (!isTextField(element)) return;
+    if (!shiftKey) {
+      return setSelectionRange(element, 0, 0);
+    }
+    const { selectionStart, selectionEnd, selectionDirection } = element;
+    // Shift+Home moves the focus to the start while the anchor stays put: the
+    // start of a forward selection, otherwise the end (treating "none" the way
+    // ArrowUp does).
+    const anchor =
+      selectionDirection === "forward"
+        ? (selectionStart ?? 0)
+        : (selectionEnd ?? 0);
+    setSelectionRange(element, 0, anchor, "backward");
   },

   async End(element, { shiftKey }) {
-    if (isTextField(element)) {
-      const { value, selectionStart } = element;
-      const start = shiftKey ? (selectionStart ?? 0) : value.length;
-      setSelectionRange(element, start, value.length, "forward");
-    }
+    if (!isTextField(element)) return;
+    const length = element.value.length;
+    if (!shiftKey) {
+      return setSelectionRange(element, length, length);
+    }
+    const { selectionStart, selectionEnd, selectionDirection } = element;
+    // Shift+End moves the focus to the end while the anchor stays put: the end
+    // of a backward selection, otherwise the start (treating "none" the way
+    // ArrowDown does).
+    const anchor =
+      selectionDirection === "backward"
+        ? (selectionEnd ?? 0)
+        : (selectionStart ?? 0);
+    setSelectionRange(element, anchor, length, "forward");
   },
```

This diff was derived from reading `press.ts` (not verified against the package's own suite here). Unit tests covering both directions belong in [`packages/ariakit-test/src/press.test.ts`](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.test.ts), which already runs in happy-dom. A natural follow-up is the related `ArrowLeft`/`ArrowRight` anchor-unaware simplification ([press.ts#L116-L148](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L116-L148)), where the focus endpoint can cross the anchor mid-range and flip direction; that is more involved, so scoping this fix to Home/End is reasonable. Since `@ariakit/test` is a published package and this changes observable selection values, a patch changeset is warranted:

```md
---
"@ariakit/test": patch
---

Fixed Shift+Home and Shift+End in `press` extending the selection past the anchor instead of deselecting the portion beyond it, matching real browser behavior.
```

*Found by an automated multi-agent audit of the repo at commit [`11bd4458d`](https://github.com/ariakit/ariakit/tree/11bd4458de6447b996dd882ea0c85eb801a6cd03) and verified with a real-browser Playwright reproduction (user-level interactions only).*

--- a/.changeset/300-test-shift-home-end.md
+++ b/.changeset/300-test-shift-home-end.md
@@ -1,0 +1,5 @@
+---
+"@ariakit/test": patch
+---
+
+Fixed `press` extending text selections past the anchor when pressing Shift+Home or Shift+End.

--- a/packages/ariakit-test/src/press.test.ts
+++ b/packages/ariakit-test/src/press.test.ts
@@ -1,6 +1,14 @@
 import { expect, test, vi } from "vitest";
 import { press } from "./index.ts";
 
+function getTextInput() {
+  const input = document.createElement("input");
+  input.value = "hello world";
+  document.body.append(input);
+  input.focus();
+  return input;
+}
+
 // A scripted click dispatched on a disabled control fires its listeners in jsdom
 // and real browsers; happy-dom drops it, but `dispatch` normalizes that (see
 // dispatch.ts), so these run in the default happy-dom. They guard the disabled
@@ -33,6 +41,28 @@ test("press.down does not activate a control that disables itself on Enter", asy
   await press.down.Enter();
 
   expect(onClick).not.toHaveBeenCalled();
+});
+
+test("press.Home with shiftKey keeps the anchor of a forward selection", async () => {
+  const input = getTextInput();
+  input.setSelectionRange(6, 8, "forward");
+
+  await press.Home(input, { shiftKey: true });
+
+  expect(input.selectionStart).toBe(0);
+  expect(input.selectionEnd).toBe(6);
+  expect(input.selectionDirection).toBe("backward");
+});
+
+test("press.End with shiftKey keeps the anchor of a backward selection", async () => {
+  const input = getTextInput();
+  input.setSelectionRange(4, 6, "backward");
+
+  await press.End(input, { shiftKey: true });
+
+  expect(input.selectionStart).toBe(6);
+  expect(input.selectionEnd).toBe(11);
+  expect(input.selectionDirection).toBe("forward");
 });
 
 test("press.Enter on a textarea emits an Enter keypress charCode", async () => {

--- a/packages/ariakit-test/src/press.ts
+++ b/packages/ariakit-test/src/press.ts
@@ -98,19 +98,34 @@ const keyDownMap: KeyActionMap = {
   },
 
   async Home(element, { shiftKey }) {
-    if (isTextField(element)) {
-      const { value, selectionEnd } = element;
-      const end = Math.min(value.length, shiftKey ? (selectionEnd ?? 0) : 0);
-      setSelectionRange(element, 0, end, "backward");
+    if (!isTextField(element)) return;
+
+    if (!shiftKey) {
+      return setSelectionRange(element, 0, 0);
     }
+
+    const { selectionStart, selectionEnd, selectionDirection } = element;
+    const anchor =
+      selectionDirection === "forward"
+        ? (selectionStart ?? 0)
+        : (selectionEnd ?? 0);
+    setSelectionRange(element, 0, anchor, "backward");
   },
 
   async End(element, { shiftKey }) {
-    if (isTextField(element)) {
-      const { value, selectionStart } = element;
-      const start = shiftKey ? (selectionStart ?? 0) : value.length;
-      setSelectionRange(element, start, value.length, "forward");
+    if (!isTextField(element)) return;
+
+    const length = element.value.length;
+    if (!shiftKey) {
+      return setSelectionRange(element, length, length);
     }
+
+    const { selectionStart, selectionEnd, selectionDirection } = element;
+    const anchor =
+      selectionDirection === "backward"
+        ? (selectionEnd ?? 0)
+        : (selectionStart ?? 0);
+    setSelectionRange(element, anchor, length, "forward");
   },
 
   async ArrowLeft(element, { shiftKey }) {

